### ISSUE TITLE
Potential of the classifier 

### ISSUE THREAD

**Issue by tom99763** [POSTECH-CVLab/PyTorch-StudioGAN#138]
Hi, your work is very interesiting

I do a [simple test on mnist](https://www.kaggle.com/code/tom99763/similarity-based-classifier?scriptVersionId=91987582) 

I think this loss has potential to extend to hard recognition task, do you plain to do that?

**Comment by MINGUKKANG** [comment]
It is very intesresting:)

When I developed D2D-CE loss, I tested D2D-CE loss for a fine-grained image retrival task using CUB200 dataset. 

I remenber that D2D-CE outperformed one of the cutting-edge metric learning losses: Proxy-Anchor in some cases. So, I think D2D-CE has potential to extend other hard recognition task.

**Comment by tom99763** [comment]
I use your optimal value formulation in the paper I found the distribution of optimal loss shwon in below

![download](https://user-images.githubusercontent.com/69715105/161432119-71231c00-4031-4091-a9a6-fd940ca7f66e.png)

I think the statistic method is an potential tool to fast tune the hyperparameters  


Code : 

`
def optimal_sampling_distribution(nclass=8,batch_size=16,nsamples=10000):
    samples=[]
    for i in range(nsamples):
        x=tf.experimental.numpy.random.randint(0,nclass,size=batch_size)
        x=tf.cast(x,'float32')
        optimal_loss=tf.reduce_mean(tf.math.log(1+x))
        samples.append(optimal_loss)
    return tf.concat(samples,axis=0)
`

**Comment by tom99763** [comment]
Here is my class distribuion from training dataset : 

![download](https://user-images.githubusercontent.com/69715105/161434423-d4d0a371-5765-4c10-9cbb-e74405591a8c.png)

And I sample batches from multinomial distribuion which parameters is based on the class distribuion and use the optimal loss formulation I get the distribuion of optimal loss

![download](https://user-images.githubusercontent.com/69715105/161434492-28af05d7-35d5-4281-914d-f96d68c142e5.png)

**Comment by MINGUKKANG** [comment]
Oh nice work!

Can you explain "I use your optimal value formulation in the paper I found the distribution of optimal loss shwon in below" more specifically ?

I don't know what that sentence means.

Thank you.

**Comment by tom99763** [comment]
> Oh nice work!
> 
> Can you explain the sentence that "I use your optimal value formulation in the paper I found the distribution of optimal loss shwon in below" more specifically ?
> 
> I don't know what that sentence means.
> 
> Thank you.

The equation A6 in your paper, when model is fitted perfectly on the training set, the D2Dloss is only related to number of flase negative samples, that's the optimal loss

**Comment by MINGUKKANG** [comment]
I've been a little busy these days, so I'm sorry for the late reply T.T

In my paper, the optimal loss of D2D-CE is as follows:

![image](https://user-images.githubusercontent.com/30516904/161950471-4b3f7503-b134-4703-baff-b45356233317.png)

where, N(i) is the set of indices that point locations of the negative samples whose labels are different from the reference label vyi in the mini-batch. So, |N(i)|  is the number of instances which have the same class label with the reference sample i.

However, when I interpret your optimal loss implementation ``optimal_loss=tf.reduce_mean(tf.math.log(1+x))``, I am very confused because you input label itself instead of |N(i)|.

Could you explain the reason for this?

Thank you.

**Comment by tom99763** [comment]
Here is my updated notebook https://www.kaggle.com/code/tom99763/sampling-distribuion-of-optimal-loss

The way I compute this optimal loss is:

Compute prior class probrability p1,...pC based on your training data which means 
Iteration 50000 times: 
    1.  sample a batch of label y1,....yN~Multinomial(n=1, p1,...pC) //1 trail and C classes
    2.  compute optimal loss based on y1,....yN

By continous sampling statistics, we get the sampling distribuion of optimal loss



> I've been a little busy these days, so I'm sorry for the late reply T.T
> 
> In my paper, the optimal loss of D2D-CE is as follows:
> 
> ![image](https://user-images.githubusercontent.com/30516904/161950471-4b3f7503-b134-4703-baff-b45356233317.png)
> 
> where, N(i) is the set of indices that point locations of the negative samples whose labels are different from the reference label vyi in the mini-batch. So, |N(i)| is the number of instances which have the same class label with the reference sample i.
> 
> However, when I interpret your optimal loss implementation `optimal_loss=tf.reduce_mean(tf.math.log(1+x))`, I am very confused because you input label itself instead of |N(i)|.
> 
> Could you explain the reason for this?
> 
> Thank you.

**Comment by tom99763** [comment]
I'm going to public my paper, and I use your idea in my paper but the task I do is image -to-image translation
